core-backend/common/src/memory.rs

```diff
@@ -0,0 +1,214 @@
+/// let read2 = manager.new_read_as::<u128>(100);
+/// let write1 = manager.new_write_as::<usize>(190);
+///
+/// // First call of read or write interface leads to pre-processing of
+/// // all already registered memory accesses, and clear `self.reads` and `self.writes`.
+/// let value_u128 = manager.read_as(read2).unwrap();
+///
+/// // Next calls do not lead to access pre-processing.
+/// let value1 = manager.read().unwrap();
+/// manager.write_as(write1, 111).unwrap();
+/// ```
+#[derive(Debug)]
+pub struct MemoryAccessManager<E: Ext> {
+    reads: Vec<(u32, u32)>,
+    writes: Vec<(u32, u32)>,
+    _phantom: PhantomData<E>,
+}
+
+impl<E: Ext> Default for MemoryAccessManager<E> {
+    fn default() -> Self {
+        Self {
+            reads: Vec::new(),
+            writes: Vec::new(),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<E: Ext> MemoryAccessManager<E> {
+    /// Register new read access.
+    pub fn new_read(&mut self, ptr: u32, size: u32) -> WasmMemoryRead {
+        self.reads.push((ptr, size));
+        WasmMemoryRead { ptr, size }
+    }
+    /// Register new read static size type access.
+    pub fn new_read_as<T: Sized>(&mut self, ptr: u32) -> WasmMemoryReadAs<T> {
+        self.reads.push((ptr, size_of::<T>() as u32));
+        WasmMemoryReadAs {
+            ptr,
+            _phantom: PhantomData,
+        }
+    }
+    /// Register new read decoded type access.
+    pub fn new_read_decoded<T: Decode + MaxEncodedLen>(
+        &mut self,
+        ptr: u32,
+    ) -> WasmMemoryReadDecoded<T> {
+        self.reads.push((ptr, T::max_encoded_len() as u32));
+        WasmMemoryReadDecoded {
+            ptr,
+            _phantom: PhantomData,
+        }
+    }
+    /// Register new write access.
+    pub fn new_write(&mut self, ptr: u32, size: u32) -> WasmMemoryWrite {
+        self.writes.push((ptr, size));
+        WasmMemoryWrite { ptr, size }
+    }
+    /// Register new write static size access.
+    pub fn new_write_as<T: Sized>(&mut self, ptr: u32) -> WasmMemoryWriteAs<T> {
+        self.writes.push((ptr, size_of::<T>() as u32));
+        WasmMemoryWriteAs {
+            ptr,
+            _phantom: PhantomData,
+        }
+    }
+    /// Call pre-processing of registered memory accesses. Clear `self.reads` and `self.writes`.
+    fn pre_process_memory_accesses(&mut self) -> Result<(), MemoryAccessError> {
+        if self.reads.is_empty() && self.writes.is_empty() {
+            return Ok(());
+        }
+        E::pre_process_memory_accesses(&self.reads, &self.writes)
+            .map_err(|_| MemoryError::OutOfBounds)?;
+        self.reads.clear();
+        self.writes.clear();
+        Ok(())
+    }
+    fn read_into_buff<M: Memory>(
+        &mut self,
+        memory: &M,
+        ptr: u32,
+        buff: &mut [u8],
+    ) -> Result<(), MemoryAccessError> {
+        self.pre_process_memory_accesses()?;
+        memory.read(ptr, buff).map_err(Into::into)
+    }
+    pub fn read<M: Memory>(
```

Comment:
Please provide doc-comments for the remaining public methods. Frankly speaking I would also ask to separate methods with empty lines for better readability.

Reply:
Done